src/edu/washington/escience/myria/operator/agg/StreamingAggregate.java

```diff
@@ -0,0 +1,199 @@
+   *
+   * @throws DbException if any error occurs.
+   * @return result tuple batch
+   */
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    if (child.eos()) {
+      return resultBuffer.popAny();
+    }
+    if (tb == null) {
+      tb = child.nextReady();
+      row = 0;
+    }
+    while (tb != null) {
+      while (row < tb.numTuples()) {
+        if (curGroupKey == null) {
+          /* First time accessing this tb, no aggregation performed previously. */
+          // store current group key as a tuple
+          curGroupKey = new Tuple(groupSchema);
+          for (int gKey = 0; gKey < gFields.length; ++gKey) {
+            TupleUtils.copyValue(tb, gFields[gKey], row, curGroupKey, gKey);
+          }
+        } else if (!TupleUtils.tupleEquals(tb, gFields, row, curGroupKey, gRange, 0)) {
+          /* Different grouping key than current one, flush current agg result to result buffer. */
+          addToResult();
+          // store current group key as a tuple
+          for (int gKey = 0; gKey < gFields.length; ++gKey) {
+            TupleUtils.copyValue(tb, gFields[gKey], row, curGroupKey, gKey);
+          }
+          aggregatorStates = AggUtils.allocateAggStates(aggregators);
+          if (resultBuffer.hasFilledTB()) {
+            return resultBuffer.popFilled();
+          }
+        }
+        // update aggregator states with current tuple
+        for (int agg = 0; agg < aggregators.length; ++agg) {
+          aggregators[agg].addRow(tb, row, aggregatorStates[agg]);
+        }
+        row++;
+      }
+      tb = child.nextReady();
+      row = 0;
+    }
+
+    /*
+     * We know that child.nextReady() has returned <code>null</code>, so we have processed all tuple we can. Child is
+     * either EOS or we have to wait for more data.
+     */
+    if (child.eos()) {
+      addToResult();
+      return resultBuffer.popAny();
+    }
+    return null;
+  }
+
+  /**
+   * Add aggregate results with previous grouping key to result buffer.
+   *
+   * @throws DbException if any error
+   */
+  private void addToResult() throws DbException {
+    int fromIndex = 0;
+    for (; fromIndex < curGroupKey.numColumns(); ++fromIndex) {
+      TupleUtils.copyValue(curGroupKey, fromIndex, 0, resultBuffer, fromIndex);
+    }
+    for (int agg = 0; agg < aggregators.length; ++agg) {
+      aggregators[agg].getResult(resultBuffer, fromIndex, aggregatorStates[agg]);
+      fromIndex += aggregators[agg].getResultSchema().numColumns();
+    }
+  }
+
+  /**
+   * The schema of the aggregate output. Grouping fields first and then aggregate fields.
+   *
+   * @return the resulting schema
+   */
+  @Override
+  protected Schema generateSchema() {
+    Operator child = getChild();
+    if (child == null) {
+      return null;
+    }
+    Schema inputSchema = child.getSchema();
+    if (inputSchema == null) {
+      return null;
+    }
+
+    groupSchema = inputSchema.getSubSchema(gFields);
+    resultSchema = groupSchema;
```

Comment:
This won't work because `groupSchema` is now aliased to `resultSchema`, so you'll be mutating both in the following code. There really should be a `Schema` copy constructor, but instead you could just do
`resultSchema = Schema.of(groupSchema.getColumnTypes(), groupSchema.getColumnNames())`.